File: server-shared/src/test/java/rikka/shizuku/server/UserServiceLifecycleTest.java
package rikka.shizuku.server;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertThrows;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import android.content.ComponentName;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.os.Bundle;
import android.os.DeadObjectException;
import android.os.Handler;
import android.os.IBinder;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.MockedStatic;
import org.mockito.Mockito;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;
import org.robolectric.shadows.ShadowBinder;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Function;

import moe.shizuku.server.IShizukuServiceConnection;
import rikka.hidden.compat.PackageManagerApis;
import rikka.shizuku.ShizukuApiConstants;
import rikka.shizuku.server.util.HandlerUtil;

/** Removal, detach and launch-cancellation behaviour of {@link UserServiceManager}. */
@RunWith(RobolectricTestRunner.class)
@Config(sdk = 34, manifest = Config.NONE)
public class UserServiceLifecycleTest {

    private static final String PACKAGE = "eu.darken.porter.probe";
    private static final int APP_ID = 10123;
    private static final int UID = APP_ID;
    private static final String DESCRIPTOR = "eu.darken.porter.probe.IProbe";

    private TestManager manager;
    private MockedStatic<PackageManagerApis> packages;
    private PackageInfo installed;
    private final AtomicReference<Runnable> pendingTimeout = new AtomicReference<>();

    private static class TestManager extends UserServiceManager {
        final List<UserServiceRecord> created = new CopyOnWriteArrayList<>();
        final List<UserServiceRecord> detached = new CopyOnWriteArrayList<>();
        final List<String> spawned = new CopyOnWriteArrayList<>();
        Function<String, String> startCmd = key -> "exit 0";

        @Override
        public String getUserServiceStartCmd(
                UserServiceRecord record, String key, String token, String packageName,
                String classname, String processNameSuffix, int callingUid, boolean use32Bits, boolean debug) {
            spawned.add(key);
            return startCmd.apply(key);
        }

        @Override
        public void onUserServiceRecordCreated(UserServiceRecord record, PackageInfo packageInfo) {
            created.add(record);
        }

        @Override
        public void onUserServiceRecordRemoved(UserServiceRecord record) {
            detached.add(record);
        }
    }

    @Before
    public void setup() {
        // A real handler would drop the timeout callback on removal, hiding the guard inside it.
        Handler handler = mock(Handler.class);
        when(handler.postDelayed(any(Runnable.class), anyLong())).thenAnswer(invocation -> {
            pendingTimeout.set(invocation.getArgument(0));
            return true;
        });
        HandlerUtil.setMainHandler(handler);

        installed = new PackageInfo();
        installed.packageName = PACKAGE;
        installed.applicationInfo = new ApplicationInfo();
        installed.applicationInfo.uid = UID;
        installed.applicationInfo.sourceDir = "/data/app/porter-probe/base.apk";
        installed.signatures = new Signature[]{new Signature("0a0b")};

        packages = Mockito.mockStatic(PackageManagerApis.class);
        packages.when(() -> PackageManagerApis.getPackageInfoNoThrow(eq(PACKAGE), anyLong(), anyInt()))
                .thenReturn(installed);

        ShadowBinder.setCallingUid(UID);
        manager = new TestManager();
    }

    @After
    public void teardown() {
        packages.close();
        ShadowBinder.reset();
    }

    private static IShizukuServiceConnection connection() {
        return new IShizukuServiceConnection.Stub() {
            @Override public void connected(IBinder service) {}
            @Override public void died() {}
        };
    }

    private Bundle options(String className) {
        Bundle options = new Bundle();
        options.putParcelable(ShizukuApiConstants.USER_SERVICE_ARG_COMPONENT, new ComponentName(PACKAGE, className));
        options.putInt(ShizukuApiConstants.USER_SERVICE_ARG_VERSION_CODE, 1);
        options.putBoolean(ShizukuApiConstants.USER_SERVICE_ARG_DAEMON, true);
        return options;
    }

    private UserServiceRecord add(String className) {
        int before = manager.created.size();
        manager.addUserService(connection(), options(className), ShizukuApiConstants.SERVER_VERSION);
        assertEquals(before + 1, manager.created.size());
        return manager.created.get(manager.created.size() - 1);
    }

    /** Publishes {@code binder} for {@code record} the way the starter's attach path does. */
    private void attach(UserServiceRecord record, IBinder binder) {
        Bundle options = new Bundle();
        options.putString(ShizukuApiConstants.USER_SERVICE_ARG_TOKEN, record.token);
        manager.attachUserService(binder, options);
    }

    private IBinder liveBinder(List<Integer> transactions) throws Exception {
        IBinder binder = mock(IBinder.class);
        when(binder.pingBinder()).thenReturn(true);
        when(binder.getInterfaceDescriptor()).thenReturn(DESCRIPTOR);
        when(binder.transact(anyInt(), any(), any(), anyInt())).thenAnswer(invocation -> {
            transactions.add(invocation.getArgument(0));
            return true;
        });
        return binder;
    }

    @SuppressWarnings("unchecked")
    private <T> T index(String name) {
        try {
            Field field = UserServiceManager.class.getDeclaredField(name);
            field.setAccessible(true);
            return (T) field.get(manager);
        } catch (ReflectiveOperationException e) {
            throw new AssertionError(e);
        }
    }

    private Map<String, UserServiceRecord> byKey() {
        return new HashMap<>(this.<Map<String, UserServiceRecord>>index("userServiceRecords"));
    }

    private Map<String, List<UserServiceRecord>> byPackage() {
        return new HashMap<>(this.<Map<String, List<UserServiceRecord>>>index("packageUserServiceRecords"));
    }

    /** The cleanup executor is single threaded, so a task queued behind ours has drained it. */
    private void drainCleanup() throws Exception {
        CountDownLatch drained = new CountDownLatch(1);
        ((java.util.concurrent.Executor) index("cleanupExecutor")).execute(drained::countDown);
        assertTrue(drained.await(5, TimeUnit.SECONDS));
    }

    @Test
    public void aSecondRemovalDoesNotDestroyTheServiceTwice() throws Exception {
        List<Integer> transactions = new CopyOnWriteArrayList<>();
        UserServiceRecord record = add("ProbeService");
        attach(record, liveBinder(transactions));

        record.removeSelf();
        record.removeSelf();
        drainCleanup();

        assertEquals(List.of(ShizukuApiConstants.USER_SERVICE_TRANSACTION_destroy), transactions);
        assertEquals(1, manager.detached.size());
    }

    @Test
    public void removingAPackageRemovesEveryRecordAndTheListItself() throws Exception {
        UserServiceRecord first = add("ProbeService");
        UserServiceRecord second = add("OtherService");

        manager.removeUserServicesForPackage(PACKAGE);
        drainCleanup();

        assertEquals(List.of(first, second), new ArrayList<>(manager.detached));
        assertTrue(byKey().isEmpty());
        assertFalse(byPackage().containsKey(PACKAGE));
    }

    @Test
    public void binderDeathPrunesBothIndexes() throws Exception {
        UserServiceRecord record = add("ProbeService");
        IBinder binder = liveBinder(new CopyOnWriteArrayList<>());
        attach(record, binder);

        org.mockito.ArgumentCaptor<IBinder.DeathRecipient> recipient =
                org.mockito.ArgumentCaptor.forClass(IBinder.DeathRecipient.class);
        Mockito.verify(binder).linkToDeath(recipient.capture(), eq(0));
        recipient.getValue().binderDied();
        drainCleanup();

        assertTrue(record.isRemoved());
        assertTrue(byKey().isEmpty());
        assertFalse(byPackage().containsKey(PACKAGE));
    }

    @Test
    public void aStartTimeoutThatFiresAfterDetachIsANoOp() throws Exception {
        UserServiceRecord record = add("ProbeService");
        Runnable timeout = pendingTimeout.get();
        assertNotNull(timeout);

        record.removeSelf();
        drainCleanup();
        assertEquals(1, manager.detached.size());

        timeout.run();

        assertEquals(1, manager.detached.size());
    }

    @Test
    public void aRemovalWhileTheStartTaskWaitsPreventsTheSpawn() throws Exception {
        CountDownLatch blocked = new CountDownLatch(1);
        CountDownLatch sentinel = new CountDownLatch(1);
        manager.startCmd = key -> {
            if (key.endsWith(":Blocker")) {
                try {
                    assertTrue(blocked.await(5, TimeUnit.SECONDS));
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            } else if (key.endsWith(":Sentinel")) {
                sentinel.countDown();
            }
            // Never reaching Runtime.exec keeps a real shell out of the test; the executor replaces
            // the worker and drains what is still queued.
            throw new IllegalStateException("no shell in tests");
        };

        add("Blocker");
        UserServiceRecord victim = add("Victim");
        add("Sentinel");

        victim.removeSelf();
        blocked.countDown();
        assertTrue(sentinel.await(5, TimeUnit.SECONDS));

        assertFalse(manager.spawned.contains(PACKAGE + ":Victim"));
        assertTrue(manager.spawned.contains(PACKAGE + ":Sentinel"));
    }

    @Test
    public void destroyAgainstADeadRemoteStillKillsTheCallbacks() throws Exception {
        UserServiceRecord record = add("ProbeService");
        IBinder binder = mock(IBinder.class);
        when(binder.pingBinder()).thenReturn(true);
        when(binder.getInterfaceDescriptor()).thenReturn(DESCRIPTOR);
        when(binder.transact(anyInt(), any(), any(), anyInt())).thenThrow(new DeadObjectException());
        attach(record, binder);

        record.removeSelf();
        drainCleanup();

        assertFalse(record.callbacks.register(connection()));
    }

    @Test
    public void aStartTimeoutRecordIsDestroyedWithoutABinder() throws Exception {
        UserServiceRecord record = add("ProbeService");

        record.removeSelf();
        drainCleanup();

        assertFalse(record.callbacks.register(connection()));
    }

    @Test
    public void tokenLivenessFollowsTheRecord() throws Exception {
        UserServiceRecord record = add("ProbeService");
        assertTrue(manager.isUserServiceTokenLive(record.token));
        assertFalse(manager.isUserServiceTokenLive("not-a-token"));
        assertFalse(manager.isUserServiceTokenLive(null));

        record.removeSelf();
        drainCleanup();

        assertFalse(manager.isUserServiceTokenLive(record.token));
    }

    @Test
    public void theAuthorisingLookupAlsoCarriesSignatures() {
        PackageInfo packageInfo = manager.ensureCallingPackageForUserService(PACKAGE, APP_ID, 0);

        assertEquals(installed, packageInfo);
        packages.verify(() -> PackageManagerApis.getPackageInfoNoThrow(
                eq(PACKAGE),
                org.mockito.ArgumentMatchers.longThat(flags ->
                        (flags & PackageManager.GET_SIGNING_CERTIFICATES) != 0
                                && (flags & 0x00002000L) != 0),
                eq(0)));
    }

    @Test
    public void theAuthorisingLookupStillRejectsAForeignPackage() {
        assertThrows(SecurityException.class,
                () -> manager.ensureCallingPackageForUserService(PACKAGE, APP_ID + 1, 0));
    }
}

File: server-shared/src/test/java/rikka/shizuku/server/UserServiceLifecycleTest.kt
package rikka.shizuku.server

import android.content.ComponentName
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.content.pm.Signature
import android.os.Bundle
import android.os.DeadObjectException
import android.os.Handler
import android.os.IBinder
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executor
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference
import moe.shizuku.server.IShizukuServiceConnection
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.ArgumentMatchers.anyLong
import org.mockito.ArgumentMatchers.eq
import org.mockito.ArgumentMatchers.longThat
import org.mockito.MockedStatic
import org.mockito.Mockito
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import org.robolectric.shadows.ShadowBinder
import rikka.hidden.compat.PackageManagerApis
import rikka.shizuku.ShizukuApiConstants
import rikka.shizuku.server.legacy.LegacyServiceConnection
import rikka.shizuku.server.util.HandlerUtil

/** Removal, detach and launch-cancellation behaviour of [UserServiceManager]. */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34], manifest = Config.NONE)
class UserServiceLifecycleTest {

    private lateinit var manager: TestManager
    private lateinit var packages: MockedStatic<PackageManagerApis>
    private lateinit var installed: PackageInfo
    private val pendingTimeout = AtomicReference<Runnable>()

    private class TestManager : UserServiceManager() {
        val created: MutableList<UserServiceRecord> = CopyOnWriteArrayList()
        val detached: MutableList<UserServiceRecord> = CopyOnWriteArrayList()
        val spawned: MutableList<String> = CopyOnWriteArrayList()
        var startCmd: (String) -> String = { "exit 0" }

        override fun getUserServiceStartCmd(
            record: UserServiceRecord,
            key: String,
            token: String,
            packageName: String,
            classname: String,
            processNameSuffix: String?,
            callingUid: Int,
            use32Bits: Boolean,
            debug: Boolean,
        ): String {
            spawned.add(key)
            return startCmd(key)
        }

        override fun onUserServiceRecordCreated(record: UserServiceRecord, packageInfo: PackageInfo) {
            created.add(record)
        }

        override fun onUserServiceRecordRemoved(record: UserServiceRecord) {
            detached.add(record)
        }
    }

    @Before
    fun setup() {
        // A real handler would drop the timeout callback on removal, hiding the guard inside it.
        val handler = mock(Handler::class.java)
        `when`(handler.postDelayed(any(Runnable::class.java), anyLong())).thenAnswer { invocation ->
            pendingTimeout.set(invocation.getArgument(0))
            true
        }
        HandlerUtil.mainHandler = handler

        installed = PackageInfo()
        installed.packageName = PACKAGE
        installed.applicationInfo = ApplicationInfo()
        installed.applicationInfo!!.uid = UID
        installed.applicationInfo!!.sourceDir = "/data/app/porter-probe/base.apk"
        installed.signatures = arrayOf(Signature("0a0b"))

        packages = Mockito.mockStatic(PackageManagerApis::class.java)
        packages.`when`<PackageInfo> { PackageManagerApis.getPackageInfoNoThrow(eq(PACKAGE), anyLong(), anyInt()) }
            .thenReturn(installed)

        ShadowBinder.setCallingUid(UID)
        manager = TestManager()
    }

    @After
    fun teardown() {
        packages.close()
        ShadowBinder.reset()
    }

    private fun connection(): IShizukuServiceConnection = object : IShizukuServiceConnection.Stub() {
        override fun connected(service: IBinder?) {}
        override fun died() {}
    }

    private fun options(className: String): Bundle {
        val options = Bundle()
        options.putParcelable(ShizukuApiConstants.USER_SERVICE_ARG_COMPONENT, ComponentName(PACKAGE, className))
        options.putInt(ShizukuApiConstants.USER_SERVICE_ARG_VERSION_CODE, 1)
        options.putBoolean(ShizukuApiConstants.USER_SERVICE_ARG_DAEMON, true)
        return options
    }

    private fun add(className: String): UserServiceRecord {
        val before = manager.created.size
        manager.addUserService(connection(), options(className), ShizukuApiConstants.SERVER_VERSION)
        assertEquals(before + 1, manager.created.size)
        return manager.created[manager.created.size - 1]
    }

    /** Publishes [binder] for [record] the way the starter's attach path does. */
    private fun attach(record: UserServiceRecord, binder: IBinder) {
        val options = Bundle()
        options.putString(ShizukuApiConstants.USER_SERVICE_ARG_TOKEN, record.token)
        manager.attachUserService(binder, options)
    }

    private fun liveBinder(transactions: MutableList<Int>): IBinder {
        val binder = mock(IBinder::class.java)
        `when`(binder.pingBinder()).thenReturn(true)
        `when`(binder.interfaceDescriptor).thenReturn(DESCRIPTOR)
        `when`(binder.transact(anyInt(), any(), any(), anyInt())).thenAnswer { invocation ->
            transactions.add(invocation.getArgument(0))
            true
        }
        return binder
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> index(name: String): T = try {
        val field = UserServiceManager::class.java.getDeclaredField(name)
        field.isAccessible = true
        field.get(manager) as T
    } catch (e: ReflectiveOperationException) {
        throw AssertionError(e)
    }

    private fun byKey(): Map<String, UserServiceRecord> = HashMap(index<Map<String, UserServiceRecord>>("userServiceRecords"))

    private fun byPackage(): Map<String, List<UserServiceRecord>> =
        HashMap(index<Map<String, List<UserServiceRecord>>>("packageUserServiceRecords"))

    /** The cleanup executor is single threaded, so a task queued behind ours has drained it. */
    private fun drainCleanup() {
        val drained = CountDownLatch(1)
        index<Executor>("cleanupExecutor").execute { drained.countDown() }
        assertTrue(drained.await(5, TimeUnit.SECONDS))
    }

    @Test
    fun aSecondRemovalDoesNotDestroyTheServiceTwice() {
        val transactions = CopyOnWriteArrayList<Int>()
        val record = add("ProbeService")
        attach(record, liveBinder(transactions))

        record.removeSelf()
        record.removeSelf()
        drainCleanup()

        assertEquals(listOf(ShizukuApiConstants.USER_SERVICE_TRANSACTION_destroy), transactions)
        assertEquals(1, manager.detached.size)
    }

    @Test
    fun removingAPackageRemovesEveryRecordAndTheListItself() {
        val first = add("ProbeService")
        val second = add("OtherService")

        manager.removeUserServicesForPackage(PACKAGE)
        drainCleanup()

        assertEquals(listOf(first, second), ArrayList(manager.detached))
        assertTrue(byKey().isEmpty())
        assertFalse(byPackage().containsKey(PACKAGE))
    }

    @Test
    fun binderDeathPrunesBothIndexes() {
        val record = add("ProbeService")
        val binder = liveBinder(CopyOnWriteArrayList())
        attach(record, binder)

        val recipient = ArgumentCaptor.forClass(IBinder.DeathRecipient::class.java)
        Mockito.verify(binder).linkToDeath(recipient.capture(), eq(0))
        recipient.value.binderDied()
        drainCleanup()

        assertTrue(record.isRemoved)
        assertTrue(byKey().isEmpty())
        assertFalse(byPackage().containsKey(PACKAGE))
    }

    @Test
    fun aStartTimeoutThatFiresAfterDetachIsANoOp() {
        val record = add("ProbeService")
        val timeout = pendingTimeout.get()
        assertNotNull(timeout)

        record.removeSelf()
        drainCleanup()
        assertEquals(1, manager.detached.size)

        timeout.run()

        assertEquals(1, manager.detached.size)
    }

    @Test
    fun aRemovalWhileTheStartTaskWaitsPreventsTheSpawn() {
        val blocked = CountDownLatch(1)
        val sentinel = CountDownLatch(1)
        manager.startCmd = { key ->
            if (key.endsWith(":Blocker")) {
                try {
                    assertTrue(blocked.await(5, TimeUnit.SECONDS))
                } catch (e: InterruptedException) {
                    Thread.currentThread().interrupt()
                }
            } else if (key.endsWith(":Sentinel")) {
                sentinel.countDown()
            }
            // Never reaching Runtime.exec keeps a real shell out of the test; the executor replaces
            // the worker and drains what is still queued.
            throw IllegalStateException("no shell in tests")
        }

        add("Blocker")
        val victim = add("Victim")
        add("Sentinel")

        victim.removeSelf()
        blocked.countDown()
        assertTrue(sentinel.await(5, TimeUnit.SECONDS))

        assertFalse(manager.spawned.contains("$USER_ID:$PACKAGE:Victim"))
        assertTrue(manager.spawned.contains("$USER_ID:$PACKAGE:Sentinel"))
    }

    @Test
    fun destroyAgainstADeadRemoteStillKillsTheCallbacks() {
        val record = add("ProbeService")
        val binder = mock(IBinder::class.java)
        `when`(binder.pingBinder()).thenReturn(true)
        `when`(binder.interfaceDescriptor).thenReturn(DESCRIPTOR)
        `when`(binder.transact(anyInt(), any(), any(), anyInt())).thenThrow(DeadObjectException())
        attach(record, binder)

        record.removeSelf()
        drainCleanup()

        assertFalse(record.callbacks.register(LegacyServiceConnection(connection())))
    }

    @Test
    fun aStartTimeoutRecordIsDestroyedWithoutABinder() {
        val record = add("ProbeService")

        record.removeSelf()
        drainCleanup()

        assertFalse(record.callbacks.register(LegacyServiceConnection(connection())))
    }

    @Test
    fun tokenLivenessFollowsTheRecord() {
        val record = add("ProbeService")
        assertTrue(manager.isUserServiceTokenLive(record.token))
        assertFalse(manager.isUserServiceTokenLive("not-a-token"))
        assertFalse(manager.isUserServiceTokenLive(null))

        record.removeSelf()
        drainCleanup()

        assertFalse(manager.isUserServiceTokenLive(record.token))
    }

    @Test
    fun theAuthorisingLookupAlsoCarriesSignatures() {
        val packageInfo = manager.ensureCallingPackageForUserService(PACKAGE, APP_ID, 0)

        assertEquals(installed, packageInfo)
        packages.verify {
            PackageManagerApis.getPackageInfoNoThrow(
                eq(PACKAGE),
                longThat { flags ->
                    (flags and PackageManager.GET_SIGNING_CERTIFICATES.toLong()) != 0L &&
                        (flags and 0x00002000L) != 0L
                },
                eq(0),
            )
        }
    }

    @Test
    fun theAuthorisingLookupStillRejectsAForeignPackage() {
        assertThrows(SecurityException::class.java) {
            manager.ensureCallingPackageForUserService(PACKAGE, APP_ID + 1, 0)
        }
    }

    private companion object {
        const val PACKAGE = "eu.darken.porter.probe"
        const val APP_ID = 10123
        const val UID = APP_ID
        const val USER_ID = 0
        const val DESCRIPTOR = "eu.darken.porter.probe.IProbe"
    }
}
